File: compilers/zig/src/ir/n115_ir_loop_cap_test.zig
//! N-115 — the unroll ceiling on the `--ir` path, and the panic it hid.
//!
//! `types.MAX_LOOP_COUNT` (10000) has existed in this tier since CL-BUG-088,
//! but the only thing that read it was `passes/anf_lower.zig` — the SOURCE
//! path. A loop arriving as IR was bounded by nothing.
//!
//! Two consequences, both measured against the checked-in `bounded-loop`
//! golden (`conformance/tests/bounded-loop/expected-ir.json`):
//!
//!   count=10001   accepted; emitted a 199734-hexchar (~97 KB) locking script.
//!                 Rust and Java accepted the identical input and emitted the
//!                 identical bytes (sha256 e2c1be39...), so the cross-tier hex
//!                 parity run saw three tiers agreeing and reported no
//!                 divergence. Only Go, Python and Ruby refused.
//!
//!   count=2^33    `thread N panic: integer does not fit in destination type`
//!                 — the `@intCast` to the `u32` count, exactly as the doc
//!                 comment on `types.MAX_LOOP_COUNT` predicted. A panic is not
//!                 a rejection: the process dies on a signal and renders no
//!                 verdict at all.
//!
//! One guard on the raw `i64`, ahead of the narrowing cast, closes both — and
//! makes the cast total, since everything reaching it is now in [0, 10000].
//!
//! Zig's IR loader is an error-enum channel with no message payload, so the
//! error NAME is the whole diagnostic: `LoopCountExceedsMaximum`, not another
//! `UnexpectedValueType` (the value's TYPE is fine; its MAGNITUDE is not).

const std = @import("std");
const testing = std.testing;
const json = @import("json.zig");
const types = @import("types.zig");

/// A one-method contract parameterised on the loop count alone, so every case
/// below differs from the control in exactly one field — the same discipline
/// `conformance/negatives/ir/I07-loop-count-over-max.ir.json` follows against
/// the `bounded-loop` golden.
fn irJson(allocator: std.mem.Allocator, count: i64) ![]const u8 {
    return std.fmt.allocPrint(allocator,
        \\{{"contractName":"Bounded","properties":[],"methods":[{{"name":"unlock","params":[],"isPublic":true,"body":[{{"name":"t0","value":{{"kind":"loop","count":{d},"iterVar":"i","body":[{{"name":"t1","value":{{"kind":"load_const","value":0}}}}]}}}}]}}]}}
    , .{count});
}

// The control. A probe whose control also fails proves nothing: a count
// exactly AT the limit is legal and must still load, or "rejects 10001" would
// only be evidence that the loader had stopped reading loops at all.
test "control: a loop count at MAX_LOOP_COUNT still loads" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const src = try irJson(allocator, types.MAX_LOOP_COUNT);
    const program = try json.parseANFProgram(allocator, src);
    try testing.expectEqualStrings("Bounded", program.contract_name);
}

test "rejects a loop count above MAX_LOOP_COUNT" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const src = try irJson(allocator, types.MAX_LOOP_COUNT + 1);
    try testing.expectError(
        error.LoopCountExceedsMaximum,
        json.parseANFProgram(allocator, src),
    );
}

// The panic half. 2^33 is outside u32 entirely, so before the guard this did
// not return an error at all — it aborted the process. `expectError` is a
// meaningful assertion here precisely because the pre-fix behaviour could not
// have satisfied it: a panicking loader never returns to the test.
test "a loop count outside u32 is an error, not a panic" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const src = try irJson(allocator, 8589934592); // 2^33
    try testing.expectError(
        error.LoopCountExceedsMaximum,
        json.parseANFProgram(allocator, src),
    );
}

// A negative count is likewise rejected rather than wrapping into a huge
// unsigned unroll. Go's loader has a companion "negative loop count"
// diagnostic; here the single magnitude guard covers both ends.
test "rejects a negative loop count" {
    var arena = std.heap.ArenaAllocator.init(testing.allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    const src = try irJson(allocator, -1);
    try testing.expectError(
        error.LoopCountExceedsMaximum,
        json.parseANFProgram(allocator, src),
    );
}

File: compilers/java/src/test/java/runar/compiler/passes/N115IrLoopCapTest.java
package runar.compiler.passes;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import org.junit.jupiter.api.Test;
import runar.compiler.ir.anf.AnfProgram;
import runar.compiler.ir.anf.Loop;

/**
 * N-115 — the unroll ceiling on the {@code --ir} path.
 *
 * <p>{@link Loop#MAX_LOOP_COUNT} (10000) has existed in this tier all along,
 * but the only thing that read it was {@link AnfLower} — the SOURCE path.
 * {@code AnfLoader.asInt} bounded the count to 32-bit signed range (R-086),
 * which is a strictly weaker claim: 10001 is a perfectly good {@code int}.
 *
 * <p>Measured against the checked-in {@code bounded-loop} golden: Go, Python
 * and Ruby rejected {@code count=10001}; Java, Rust and Zig accepted it and all
 * three emitted the SAME 199734-hexchar (~97 KB) script (sha256
 * {@code e2c1be39...}). Three tiers agreeing is indistinguishable from three
 * tiers being right to a gate that only diffs bytes, which is why this needed
 * a negative fixture —
 * {@code conformance/negatives/ir/I07-loop-count-over-max.ir.json}, that same
 * golden with this one field changed.
 */
class N115IrLoopCapTest {

    /**
     * A one-method contract parameterised on the loop count alone, so every
     * case below differs from the control in exactly one field.
     */
    private static String ir(long count) {
        return """
            {"contractName":"Bounded","properties":[],"methods":[
              {"name":"unlock","params":[],"isPublic":true,"body":[
                {"name":"t0","value":{"kind":"loop","count":%d,"iterVar":"i","body":[
                  {"name":"t1","value":{"kind":"load_const","value":0}}
                ]}}
              ]}
            ]}
            """.formatted(count);
    }

    /**
     * The control. A count exactly AT the limit is legal and must still load —
     * without this row, "rejects 10001" would be equally consistent with a
     * loader that had stopped accepting loops at all.
     */
    @Test
    void controlLoopCountAtTheLimitStillLoads() {
        AnfProgram p = AnfLoader.parse(ir(Loop.MAX_LOOP_COUNT));
        assertEquals("Bounded", p.contractName());
    }

    @Test
    void rejectsLoopCountOverMax() {
        RuntimeException ex = assertThrows(
            RuntimeException.class, () -> AnfLoader.parse(ir(Loop.MAX_LOOP_COUNT + 1L)));
        assertTrue(
            ex.getMessage().contains("loop count 10001 exceeding maximum 10000"),
            "diagnostic must name the count and the limit, as Go's does: " + ex.getMessage());
    }

    /**
     * The pre-existing R-086 range guard is the OUTER bound and still fires
     * first for a value no {@code int} can hold. Pinned here so a future edit
     * cannot reorder the two checks into one that silently narrows before
     * comparing.
     */
    @Test
    void anOutOf32BitRangeCountIsStillRejectedByTheRangeGuard() {
        RuntimeException ex = assertThrows(
            RuntimeException.class, () -> AnfLoader.parse(ir(8589934592L))); // 2^33
        assertTrue(
            ex.getMessage().contains("out of 32-bit signed range"), ex.getMessage());
    }
}
